### DISCUSSION TITLE
Understanding Multi-Vehicle vs. Single-Vehicle Algorithm Efficiency in Visit Allocation

### DISCUSSION THREAD

**Discussion by ibiscp** [google/or-tools#4233] [Routing (and legacy CP) questions]
Hi everyone,

I'm working on a system where drivers select sets of visits from a shared pool with penalty for dropped visits. I've observed that running the allocation algorithm for multiple vehicles at once (e.g., 4 vehicles) yields better results than running the algorithm individually for a single vehicle four separate times.

I want to generate it on demand because I don't know the capacity of each vehicle, and also because the pool of visits can also be updated.

Could anyone explain why the algorithm performs better in the multi-vehicle scenario? Additionally, is there a way to achieve the same efficiency when the algorithm runs sequentially for one vehicle at a time?

**Comment by sschnug** [comment]
Sounds to me like an expected trade-off between greedy-sequential (N=1) and global-optimization (N=N).

I would (carefully) claim, that with a perfect oracle, this is a monotonic function in terms of solution-quality (worst to best).

With incomplete solvers, like the routing-solver, things are less predictable, but (mini-)batching some visits allows more reasoning about the potential (shared-resource )conflicts: less narrow / fixed decision-options; more freedom.

Edit: Ah... vehicles, not visits. Well i guess, the argument would be the same, although i would expect less impact.

**Reply by ibiscp** [reply]
Implementing batching involves the additional need to optimize for the number of vehicles. Consider a scenario where we have a pool of 1000 nodes, and we batch 10 nodes at a time. The optimal solution will vary based on the number of vehicles used, as there is an ideal number of vehicles that minimizes the total distance traveled.

Another complication arises towards the end of the process. For the last batch of 10 nodes, if we maintain the same number of vehicles (say 10), each vehicle would only manage one node. This scenario would require either reducing the number of vehicles for the final batch or adjusting the distribution strategy.

Ideally, I would prefer not to have the number of vehicles as a variable in this equation, as it complicates the optimization process.

**Reply by sschnug** [reply]
I just explained the potential origin of the effect.

You seem to be more interested in a meta-algorithm but we don't know really what requirements are there. It seems, "online"-capability / missing configuration-information is the core reason behind this decomposition.

But even then... nobody stops you from doing LNS-like post-optimization in-between your fine-grained steps.
- single(1) -> 1 tours
- single(1) -> 1 tours
- optional batched step: (batch(4) -> 3 tours)
- LNS(tours touched = tour0, tour2) -> 2 tours (got rid of one tour by reordering things) | tune LNS neighborhood size such that not too big, e.g. touch 2 vs. ALL tours
- ...

If you really have no idea about vehicle-capas, single-opt is the easiest to provide guarantees (guessing and correcting vehicle-capas is a bit more complex) and doing every N steps some "more-global" post-opt can help to improve things while basically only assuming, that you don't have to commit to execution before reaching post-opt results (or you might need to freeze parts not mutable).
